### PR TITLE
[Neo VM Style] Add comments and document to reference counter

### PR DESCRIPTION
**Pull request by Jim8y**
# Description

Neo VM reference counter is a very important component, but it is very complex and is hard to understand. Making it actually **UNMAINTAINABLE**. Thus having this pr that adds detailed explaination and comments to all its related code. 

Fixes # None

## Type of change

- [ ] Optimization (the change is only an optimization)
- [x] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

**Test Configuration**:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/Neo.VM/Types/Array.cs [hunk 1]
```diff
@@ -142,5 +142,36 @@ public void Reverse()
...
+                        {
+                            visited.Add(subItem);
+                            stack.Push(subItem);
+                        }
+                    }
+                }
+            }
+            return false;
+        }
+
```
Comment (Jim8y) [inline src/Neo.VM/Types/Array.cs:175]:
```suggestion
```

File: src/Neo.VM/ReferenceCounter.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    - If there are no cached SCCs, it recomputes them using Tarjan's algorithm. This involves creating a `Tarjan` object, passing the `tracked_items` to it, and invoking the algorithm to get the SCCs.
+
+3. **Reset Tarjan-related Fields**:
+    - It resets the Tarjan-related fields (`DFN`, `LowLink`, `OnStack`) of all tracked items to prepare for processing SCCs.
+
+4. **Process Each SCC**:
+    - It iterates through each SCC (component) in the cached components list. For each component, it checks if any item is still on the stack by looking at its `StackReferences` or if any of its parent items are on the stack.
+
+
+
```
Comment (Jim8y) [inline src/Neo.VM/ReferenceCounter.md:226]:
```suggestion
```

**Review by shargon — COMMENTED** [review]
I think that we should wait for https://github.com/neo-project/neo/issues/3328 before merge it, because of conflicts

**Comment by Jim8y** [comment]
> I think that we should wait for #3328 before merge it, because of conflicts

should be fine, its just comments.

**Review by shargon — APPROVED** [review]

File: src/Neo.VM/ReferenceCounter.cs [hunk 3]
```diff
@@ -22,86 +22,180 @@ namespace Neo.VM
...
+        /// </summary>
+        /// <param name="item">The item to add a reference to.</param>
+        /// <param name="parent">The parent compound type.</param>
         internal void AddReference(StackItem item, CompoundType parent)
         {
-            references_count++;
+            // Increment the reference count.
+            _referencesCount++;
+
+            // If the item doesn't need to be tracked, return early.
```
Comment (vncoelho) [inline src/Neo.VM/ReferenceCounter.cs:79]: return earlier?

File: src/Neo.VM/ReferenceCounter.cs [hunk 4]
```diff
@@ -22,86 +22,180 @@ namespace Neo.VM
...
+            _trackedItems.Add(item);
+
+            // Initialize the ObjectReferences dictionary if it is null.
+            item.ObjectReferences ??= new Dictionary<CompoundType, StackItem.ObjectReferenceEntry>(ReferenceEqualityComparer.Instance);
+
+            // Add the parent to the item's ObjectReferences dictionary and increment its reference count.
             if (!item.ObjectReferences.TryGetValue(parent, out var pEntry))
             {
-                pEntry = new(parent);
+                pEntry = new StackItem.ObjectReferenceEntry(parent);
```
Comment (vncoelho) [inline src/Neo.VM/ReferenceCounter.cs:95]: Can you revert this change from here?

Comment (Jim8y) [reply src/Neo.VM/ReferenceCounter.cs:95]: Problems like this are very minor, not worth a pr to fix, but if we don't fix it, it will always be this unclear syntax.

File: src/Neo.VM/ReferenceCounter.cs (hunk 3 above)
Comment (vncoelho) [reply src/Neo.VM/ReferenceCounter.cs:79]: forget, wrong from my side

File: src/Neo.VM/ReferenceCounter.cs [hunk 5]
```diff
@@ -22,86 +22,180 @@ namespace Neo.VM
...
+                // Clear the zero_referred set since we are going to process all of them.
+                _zeroReferred.Clear();
+
+                // If cached components are null, we need to recompute the strongly connected components (SCCs).
+                if (_cachedComponents is null)
                 {
-                    //Tarjan<StackItem> tarjan = new(tracked_items.Where(p => p.StackReferences == 0));
-                    Tarjan tarjan = new(tracked_items);
-                    cached_components = tarjan.Invoke();
+                    // Create a new Tarjan object and invoke it to find all SCCs in the tracked_items graph.
```
Comment (vncoelho) [inline src/Neo.VM/ReferenceCounter.cs:172]: Next PR could be comment Tarjan class

File: src/Neo.VM/ReferenceCounter.cs [hunk 6]
```diff
@@ -22,86 +22,180 @@ namespace Neo.VM
...
+                // Process each SCC in the cached_components list.
+                for (var node = _cachedComponents.First; node != null;)
                 {
                     var component = node.Value;
                     bool on_stack = false;
+
+                    // Check if any item in the SCC is still on the stack.
                     foreach (StackItem item in component)
                     {
+                        // An item is considered 'on stack' if it has stack references or if its parent items are still on stack.
```
Comment (vncoelho) [inline src/Neo.VM/ReferenceCounter.cs:190]: This could be above as I mentioned

File: src/Neo.VM/ReferenceCounter.cs (hunk 5 above)
Comment (Jim8y) [reply src/Neo.VM/ReferenceCounter.cs:172]: You can do that, it's too complex to understand.... I just assume it works fine. LOL

File: src/Neo.VM/ReferenceCounter.cs [hunk 7]
```diff
@@ -22,86 +22,180 @@ namespace Neo.VM
...
-                    //Tarjan<StackItem> tarjan = new(tracked_items.Where(p => p.StackReferences == 0));
-                    Tarjan tarjan = new(tracked_items);
-                    cached_components = tarjan.Invoke();
+                    // Create a new Tarjan object and invoke it to find all SCCs in the tracked_items graph.
+                    Tarjan tarjan = new(_trackedItems);
+                    _cachedComponents = tarjan.Invoke();
                 }
-                foreach (StackItem item in tracked_items)
+
+                // Reset all tracked items' Tarjan algorithm-related fields (DFN, LowLink, and OnStack).
```
[the remark this answers is not in the document]
Comment (vncoelho) [reply src/Neo.VM/ReferenceCounter.cs:177]: they are explained on Vertex.

**Review by vncoelho — COMMENTED** [review]
Looks good, @Jim8y 

There is just a minor change to be reverted I think

File: src/Neo.VM/ReferenceCounter.cs (hunk 5 above)
Comment (vncoelho) [reply src/Neo.VM/ReferenceCounter.cs:172]: EHUiaheuahea, I will ask @igormcoelho

Comment (vncoelho) [reply src/Neo.VM/ReferenceCounter.cs:172]: He knows this Tarjan algorithm very well, he teaches that.

File: src/Neo.VM/ReferenceCounter.cs (hunk 4 above)
Comment (vncoelho) [reply src/Neo.VM/ReferenceCounter.cs:95]: Fix in another PR, it is better.
Let just comments here

Even the underscore could be moved to another PR.

File: src/Neo.VM/ReferenceCounter.cs (hunk 5 above)
Comment (Jim8y) [reply src/Neo.VM/ReferenceCounter.cs:172]: Awesome!!!!! Ask him the meaning of every single letter in that algorithm and comment them.

**Review by cschuchardt88 — APPROVED** [review]
